src/status_im/transport/shh.cljs

```diff
@@ -0,0 +1,161 @@
+(ns status-im.transport.shh
+  (:require [taoensso.timbre :as log]
+            [re-frame.core :as re-frame]
+            [status-im.transport.utils :as transport.utils]
+            [status-im.transport.message.transit :as transit]))
+
+(defn get-new-key-pair [{:keys [web3 on-success on-error]}]
+  (if web3
+    (.. web3
+        -shh
+        (newKeyPair (fn [err resp]
+                      (if-not err
+                        (on-success resp)
+                        (on-error err)))))
+    (on-error "web3 not available.")))
+
+(re-frame/reg-fx
+  :shh/get-new-key-pair
+  (fn [{:keys [web3 success-event error-event]}]
+    (get-new-key-pair {:web3       web3
+                       :on-success #(re-frame/dispatch [success-event %])
+                       :on-error   #(re-frame/dispatch [error-event %])})))
+
+(defn get-public-key [{:keys [web3 key-pair-id on-success on-error]}]
+  (if (and web3 key-pair-id)
+    (.. web3
+        -shh
+        (getPublicKey key-pair-id (fn [err resp]
+                                    (if-not err
+                                      (on-success resp)
+                                      (on-error err)))))
+    (on-error "web3 or key-pair id not available.")))
+
+(re-frame/reg-fx
+  :shh/get-public-key
+  (fn [{:keys [web3 key-pair-id success-event error-event]}]
+    (get-public-key {:web3        web3
+                     :key-pair-id key-pair-id
+                     :on-success  #(re-frame/dispatch [success-event %])
+                     :on-error    #(re-frame/dispatch [error-event %])})))
+
+(defn generate-sym-key-from-password
+  [{:keys [web3 password on-success on-error]}]
+  (.. web3
+      -shh
+      (generateSymKeyFromPassword password (fn [err resp]
+                                             (if-not err
+                                               (on-success resp)
+                                               (on-error err))))))
+
+(defn post-message
+  [{:keys [web3 whisper-message on-success on-error]}]
+  (.. web3
+      -shh
+      (post (clj->js whisper-message) (fn [err resp]
+                                        (if-not err
+                                          (on-success resp)
+                                          (on-error err))))))
+(re-frame/reg-fx
+  :shh/post
+  (fn [{:keys [web3 message success-event error-event]
+        :or {error-event :protocol/send-status-message-error}}]
+    (post-message {:web3       web3
+                   :whisper-message (update message :payload (comp transport.utils/from-utf8
+                                                                   transit/serialize))
+                   :on-success (if success-event
+                                 #(re-frame/dispatch success-event)
+                                 #(log/debug :shh/post-success))
+                   :on-error   #(re-frame/dispatch [error-event %])})))
+
+(re-frame/reg-fx
+  :shh/multi-post
```

Comment:
Multipass! New thing. I don't understand why shh/post doesn't have `recipient` but this has multiple, nor why this fn would be used.

Reply:
This fn has recipients because it's a vector of map with public-key and topic keys. post has public-key and topic keys at the top level of the args map. This fn is used to send messages to multiple recipients when you can't send it on a topic. It is used for renewing keys in a private group chat, because if someone leaves/join, you want to change the symmetric key but you can't do that in the group topic with the old key otherwise leavers can still eavesdrop / joiners can read past history.

Reply:
Ok nice, good docstring

Reply:
yeah I will pretty much put that as docstring